pinot-core/src/main/java/org/apache/pinot/core/plan/TransformPlanNode.java

```diff
@@ -79,16 +77,13 @@ private void extractProjectionColumns(TransformExpressionTree expression, Set<St
   /**
    * Helper method to set the max number of docs to return for selection queries
    */
-  private void setMaxDocsForSelection(BrokerRequest brokerRequest) {
-    if (!brokerRequest.isSetAggregationsInfo()) {
-      Selection selection = brokerRequest.getSelections();
-
-      // Update MaxDocPerNextCall
-      if (selection.getSize() > 0) {
-        List<SelectionSort> sortSequence = selection.getSelectionSortSequence();
-        if (sortSequence == null) {
-          // For selection only queries, select minimum number of documents
-          _maxDocPerNextCall = Math.min(selection.getSize(), _maxDocPerNextCall);
+  private void setMaxDocsForSelection(QueryContext queryContext) {
+    if (!QueryContextUtils.isAggregationQuery(queryContext)) {
+      // Selection queries
+      if (queryContext.getLimit() > 0) {
```

Comment:
The limit should be set to the max (1, user specified limit) or default (10) if not specified. This can be done at the parse time

Reply:
This is already done on the broker side, and we allow queries with `LIMIT 0`

Reply:
I don't know why we have EmptySelectionOperator. The broker should return an empty response right away with just the schema for LIMIT 0. Anyway, it is not related to this PR.

Reply:
The LIMIT 0 query is meant to get the schema of the table. When all table have schema and the schema is cached on broker side, we can directly return on broker side.